javaslang/src/main/java/javaslang/Lazy.java

```diff
@@ -135,14 +133,11 @@ private Lazy(Supplier<? extends T> supplier) {
      */
     @Override
     public T get() {
-        // using a local var speeds up the double-check idiom by 25%, see Effective Java, Item 71
-        Supplier<? extends T> tmp = supplier;
-        if (tmp != null) {
+        if ((value == null) && (supplier != null)) {
```

Comment:
When we use synchronized, this is always OK because the synch block ensures that supplier.get() (which sets the atomic boolean) is only entered by one thread at a time:

```java
assertThat(isEvaluated.getAndSet(true)).isFalse();
```

The second assert also can't fail because value has always the correct value. The worst case that can happen is, that supplier.get() is called twice (because the second thread did not see the initialized var).

The latter case (twice initialized) is what you see for non-volatile vars without the synchronized block. Because the synchronized block is missing, the first assertion fails (supplier.get() entered by two threads and atomic boolean assertion fails).

Right?